### ISSUE TITLE
Update OSX openssl requirement to openssl@3 

### ISSUE THREAD

**Issue by crpahl** [rvm/rvm#5380]
## Description

The current openssl requirement for OSX [here](https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/osx_brew#L334) is `openssl@1.1`. The end of life date for that version is September, 11 2023. Would it be reasonable to update that requirement to `openssl@3`?

Homebrew has updated a number of its formulae recently to use `openssl@3`:

- https://github.com/Homebrew/homebrew-core/issues/134251

As a result, when we follow the security steps [here](https://rvm.io/rvm/security) (and install `gpg` using `brew install gnupg`), that will install `openssl@3` as a brew dependency. When we then run `rvm install 3.2.2` that will install `openssl@1.1`. At that point we have this:

```
brew list --versions openssl
openssl@3 3.1.2
```

```
brew list | grep openssl
openssl@1.1
openssl@3
```

When _both_ versions have been installed by brew, we get the error in my gist below. When _only_ `openssl@1.1` has been installed by brew, `rvm install 3.2.2` will work.

We've gotten around this for now (and maybe this is just always the path forward) with this:

`rvm install 3.1.4 --with-openssl-dir=`brew --prefix openssl@3``

## Steps to reproduce

- `brew install gnupg`
- `rvm install 3.2.2`

## Expected behavior

Successful install of Ruby `3.2.2` after running `rvm install 3.2.2`

## Actual behavior

* [rvm install 3.2.2](https://gist.github.com/crpahl/c7b8d991879fab2afd1dbd7f351f8e84)
* [make.log](https://gist.github.com/crpahl/76107ccdda920f1c7625abfeed310456) - only for issues with `install`

## Environment info

```
ruby-3.1.4:

  system:
    uname:        "Darwin Clintons-MacBook-Pro.local 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000 arm64 arm Darwin"
    name:         "OSX"
    version:      "13.3"
    architecture: "arm64"
    bash:         "/opt/homebrew/bin/bash => GNU bash, version 5.2.15(1)-release (aarch64-apple-darwin22.1.0)"
    zsh:          "/bin/zsh => zsh 5.9 (x86_64-apple-darwin22.0)"
    remote_path:  "osx/13.3/arm64"
    xcode:        "14.3.1"

  rvm:
    version:      "1.29.12 (latest)"
    updated:      "1 hour 55 minutes 9 seconds ago"
    path:         "/Users/clintonpahl/.rvm"
    autolibs:     "[4] Allow RVM to use package manager if found, install missing dependencies, install package manager (only OS X)."

  ruby:
    interpreter:  "ruby"
    version:      "3.1.4p223"
    date:         "2023-03-30"
    platform:     "arm64-darwin22"
    patchlevel:   "2023-03-30 revision 957bb7cb81"
    full_version: "ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [arm64-darwin22]"

  homes:
    gem:          "/Users/clintonpahl/.rvm/gems/ruby-3.1.4"
    ruby:         "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4"

  binaries:
    ruby:         "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/bin/ruby"
    irb:          "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/bin/irb"
    gem:          "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/bin/gem"
    rake:         "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/bin/rake"

  environment:
    PATH:         "/Users/clintonpahl/.rvm/gems/ruby-3.1.4/bin:/Users/clintonpahl/.rvm/gems/ruby-3.1.4@global/bin:/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/bin:/Users/clintonpahl/.rvm/bin:/opt/homebrew/opt/coreutils/libexec/gnubin:/opt/homebrew/opt/findutils/libexec/gnubin:/opt/homebrew/opt/gawk/libexec/gnubin:/Users/clintonpahl/Library/Caches/fnm_multishells/51973_1691083789121/bin:/Users/clintonpahl/.pyenv/shims:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/clintonpahl/workspace/aws-tooling/bin"
    GEM_HOME:     "/Users/clintonpahl/.rvm/gems/ruby-3.1.4"
    GEM_PATH:     "/Users/clintonpahl/.rvm/gems/ruby-3.1.4:/Users/clintonpahl/.rvm/gems/ruby-3.1.4@global"
    MY_RUBY_HOME: "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4"
    IRBRC:        "/Users/clintonpahl/.rvm/rubies/ruby-3.1.4/.irbrc"
    RUBYOPT:      ""
    gemset:       ""
```

**Comment by acdesouza** [comment]
The [requirements script for osx using brew](https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/osx_brew#L317) could be related?

The [default openssl Is the Homebrew package openssl@1.1](https://github.com/rvm/rvm/blob/master/scripts/functions/requirements/osx_brew#L333-L334)

**Comment by BrianHawley** [comment]
Ruby 3.2.x and before won't compile with openssl 3.1, and rvm can't change that.

Specify a different openssl version that is supported. For example:
```
rvm install 3.2.2 --with-openssl-dir=$(brew --prefix openssl@1.1)
```

Or if you want a more recent version, you can use openssl 3.0:
```
brew install openssl@3.0
rvm install 3.2.2 --with-openssl-dir=$(brew --prefix openssl@3.0)
```

Or you could use a version of ruby that supports openssl 3.1, like this:
```
rvm install 3.3.0-preview2 --with-openssl-dir=$(brew --prefix openssl@3)
```

**Comment by jedrekdomanski** [comment]
I have the same problem. Every ruby installation fails with an error pointing to the outdated openssl and I need to manually point to the newer version of openssl. It would be nice if RVM could use the latest version automatically.

**Comment by jshier** [comment]
@jedrekdomanski How have you been pointing RVM to the new OpenSSL?

`brew doctor` is now recommending people delete their installs of OpenSSL 1.1, so anyone who takes that advice (me) will have their existing RVM Ruby installs broken and any new installs will fail. So this is now a breaking issue that needs a fix sooner rather than later.

**Comment by jedrekdomanski** [comment]
@jshier I've been installing Ruby by passing the extra flag `--with-openssl-dir`, for instance:
```
rvm install 2.7.4 --with-openssl-dir=$(brew --prefix openssl@3)
```
`brew` refuses to uninstall older version of openssl due to dependencies
```
brew uninstall openssl@1.1
Error: Refusing to uninstall /opt/homebrew/Cellar/openssl@1.1/1.1.1w
because it is required by cffi, docutils, dotnet, openconnect, pycparser and vpn-slice, which are currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies openssl@1.1
```
so I don't know how brew can recommend uninstalling openssl version 1.1

**Comment by jshier** [comment]
Thanks, I just found that myself. I see others also running with `--reconfigure`, which is probably a good idea when you have new developer tools and OS versions installed.

**Comment by jedrekdomanski** [comment]
Now I can't even install `rvm install ruby-2.7.2 --with-openssl-dir=$(brew --prefix openssl@3)` so can't install any Ruby version

**Comment by crpahl** [comment]
@jedrekdomanski Can you try `rvm install ruby-2.7.2 --with-openssl-dir=$(brew --prefix openssl@1.1)`? That version of Ruby needs OpenSSL 1.1

**Comment by jedrekdomanski** [comment]
> @jedrekdomanski Can you try `rvm install ruby-2.7.2 --with-openssl-dir=$(brew --prefix openssl@1.1)`? That version of Ruby needs OpenSSL 1.1

It didn't work.

**Comment by crpahl** [comment]
Hmm, we're on the latest version of Sonoma and haven't run into an issue, but we're on Ruby `3.2.2` and are able to install it with:

```
rvm install 3.2.2 --with-openssl-dir=`brew --prefix openssl@3`
```

That won't work for you though since you need OpenSSL `1.1` for Ruby `2.7.2`. When it failed, it should give you a log file that can look at to see the specific errors. Curious what that says if you can look at that.

**Comment by jedrekdomanski** [comment]
It's "C" type unintelligible and meaningless gobbledygook :) See attached.

[make.log](https://github.com/rvm/rvm/files/13821744/make.log)

**Comment by crpahl** [comment]
Thanks! Yeah I was hoping an obvious error that I had see myself before would stand out to me, but I'm not in this area enough myself and didn't see anything. I think we need help from another subject matter expert at this point 😞

Doing something like `rvm install ruby-2.7.2 --with-openssl-dir=$(brew --prefix openssl@1.1)` has worked for us in the past, but like I said, we're on the latest Ruby version now and haven't needed to do that in a while. When I get a moment, I'll try installing the Ruby version you're trying to install and see if I have an issue. If I do, [this PR](https://github.com/rvm/rvm/pull/5394) I opened a while back will need to be updated...

Honestly, I might just close that PR at this point since multiple people in the rvm community are having challenges and I don't want to give false hope about a solution landing since I'm having trouble moving it along...

**Comment by crpahl** [comment]
Oh I think I know what the issue is. What does `brew list | grep openssl` give you? I've noticed that `rvm` will fail to install when there's multiple versions of `openssl` installed with brew. You'll probably see something like this:

```
brew list | grep openssl
openssl@1.1
openssl@3
```

I was able to install version `2.7.2` by uninstalling openssl 3 first:

```
brew uninstall --ignore-dependencies openssl@3
```

And then I was able to do:

```
rvm install ruby-2.7.2 --with-openssl-dir=$(brew --prefix openssl@1.1)
```

The reason you'll need `--ignore-dependencies openssl@3` is because that version of `openssl` is used in many of the [default Homebrew forumulae now](https://github.com/Homebrew/homebrew-core/issues/134251). So you'll just need to be aware that uninstalling that version of openssl may effect other things installed through Homebrew.

**Comment by vikram-a** [comment]
+1

No other solution is working for me.

**Comment by monfresh** [comment]
There are many factors that can affect Ruby installation on a Mac, but I want to focus on 2 issues from the comments that are important to understand:

1) Different Ruby versions support different OpenSSL versions. This is a common point of confusion, so I wrote this blog post that shows which OpenSSL versions are supported by Ruby versions from 1.9 to 3.3: https://www.rubyonmac.dev/openssl-versions-supported-by-ruby

2) There is rarely a good reason to use a Ruby version where the patch version (the 3rd digit) is not the latest. For example, @jedrekdomanski has been trying to install Ruby 2.7.2. However, the latest version in the 2.7.x series is 2.7.8. There are known issues with older 2.7 versions, so the first thing I would try is to see if you can install 2.7.8. If so, then the next step would be to upgrade the Ruby version in your project. Here's a step-by-step guide that explains [how and why to upgrade the Ruby version in your project](https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project).

The other important thing to realize is that there are other tools you can use to install Ruby. Try one of the other version managers, such asdf, frum, rbenv, or ruby-install. Another option is to buy my [Ruby on Mac](https://www.rubyonmac.dev/) product that is guaranteed to get you up and running in 15 minutes or less.

**Comment by crpahl** [comment]
> The other important thing to realize is that there are other tools you can use to install Ruby.

Yeah, thanks for sharing that. We're actually going to move to `asdf` ourselves. We've had no issues with it so far.

> Different Ruby versions support different OpenSSL versions

I tried addressing that in the PR I started, but I'm having trouble moving it forward and I think it still has some gaps:

https://github.com/rvm/rvm/pull/5394#discussion_r1388480994

**Comment by vikram-a** [comment]
> 1. Different Ruby versions support different OpenSSL versions. 

I thought it it RVM that is tied to OpenSSL 3 when checking requirements for Mac, but is unable to use it. But if that is the case, the OS cannot have multiple versions of OpenSSL installed. This makes it more complicated. But thanks for your reply.

**Comment by michael-reeves** [comment]
@jedrekdomanski for the problems you were having with getting rid of openssl@1, I ended up uninstalling everything installed by homebrew and reinstalling the pieces I most cared about.
```bash
brew leaves # to get a list of packages I personally installed on the system
brew leaves | xargs brew uninstall
brew list  # this showed the packages which were still installed
brew list | xargs brew uninstall
rm -rf /usr/local/etc/openssl@1.1. # configuration files were not removed during installation, so I wanted it completely gone
brew list # verified everything gone
brew doctor # verified everything was still good to go
brew cleanup
brew install <each of the packages previously listed in brew leaves>  # I also checked the dependencies of each one to make sure openssl@1 didn't come back
```

However when trying to run 
```bash
rvm reinstall 3.3.4 --with-openssl-dir=`brew --prefix openssl@3`
```
, RVM was still trying to install `openssl@1.1`.  To get around this, I grabbed the latest version of RVM (`rvm get stable`) and then edited the requirements file listed above, because the * catchall was still trying to install the old version of openssl.

So I updated line 332 of `~/.rvm/scripts/functions/requirements/osx_brew` to remove the `openssl@1.1` requirement
```bash
(*)
  brew_openssl_package="openssl"
  ;;
```

Then I was able to reinstall 3.3.4 correctly with 
```bash
rvm reinstall 3.3.4 --with-openssl-dir=`brew --prefix openssl@3
```
This might not be a fantastic solution if you're trying to install older rubies, and it will obviously have to be done again when I pull RVM, but it worked for me for the time being.

**Comment by jedrekdomanski** [comment]
> @jedrekdomanski for the problems you were having with getting rid of openssl@1, I ended up uninstalling everything installed by homebrew and reinstalling the pieces I most cared about.
> 
> ```shell
> brew leaves # to get a list of packages I personally installed on the system
> brew leaves | xargs brew uninstall
> brew list  # this showed the packages which were still installed
> brew list | xargs brew uninstall
> rm -rf /usr/local/etc/openssl@1.1. # configuration files were not removed during installation, so I wanted it completely gone
> brew list # verified everything gone
> brew doctor # verified everything was still good to go
> brew cleanup
> brew install <each of the packages previously listed in brew leaves>  # I also checked the dependencies of each one to make sure openssl@1 didn't come back
> ```
> 
> However when trying to run
> 
> ```shell
> rvm reinstall 3.3.4 --with-openssl-dir=`brew --prefix openssl@3`
> ```
> 
> , RVM was still trying to install `openssl@1.1`. To get around this, I grabbed the latest version of RVM (`rvm get stable`) and then edited the requirements file listed above, because the * catchall was still trying to install the old version of openssl.
> 
> So I updated line 332 of `~/.rvm/scripts/functions/requirements/osx_brew` to remove the `openssl@1.1` requirement
> 
> ```shell
> (*)
>   brew_openssl_package="openssl"
>   ;;
> ```
> 
> Then I was able to reinstall 3.3.4 correctly with
> 
> ```shell
> rvm reinstall 3.3.4 --with-openssl-dir=`brew --prefix openssl@3
> ```
> 
> This might not be a fantastic solution if you're trying to install older rubies, and it will obviously have to be done again when I pull RVM, but it worked for me for the time being.

I fixed it log time ago by upgrading RVM to the latest version and everything started to work correctly.

**Comment by JorenVos** [comment]
Since 24/10/2024, it is no longer to install **any** Ruby version using RVM due to the fact that openssl requirement is set to openssl@1.1 .

**Comment by workmaster2n** [comment]
I disabled rvm autolibs with `rvm autolibs disable` and then ran `rvm install ruby-3.3.5 --with-openssl-dir=$(brew --prefix openssl@3)`, which "installed" ruby 3.3.5, but I don't know if I'll run into an issue later, given what @JorenVos has pointed out and openssl@1.1 is no longer available.

**Comment by monfresh** [comment]
OpenSSL 1.1 should only be used for Ruby versions older than 3.1. Anything newer than or equal to Ruby 3.1 should be using OpenSSL 3.x. See this [OpenSSL and Ruby Compatibility Table](https://www.rubyonmac.dev/openssl-versions-supported-by-ruby).

Also, Ruby versions older than 3.1 (those that require OpenSSL 1.1) have reached end of life, so they shouldn't be used anyways. If you need those older Ruby versions for an existing project, your goal should be to update them to Ruby 3.1 at least. Here's a step-by-step guide that shows [How and Why to Upgrade the Ruby Version in Your Project](https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project).

**Comment by workmaster2n** [comment]
Sorry for the confusion @monfresh - Even when trying to install ruby-3.3.5, RVM thinks there is a dependency on `openssl 1.1`. I don't think there is an actual dependency from Ruby, I think there is a baked in dependency somewhere in the RVM install chain.

**Comment by gregawoods** [comment]
I've had to use `--with-openssl-dir=$(brew --prefix openssl@3.0)` when installing _anything_ in RVM for a long time now. I've just accepted it as a fact of life at this point. 🤔 

Just attempted to install the newly released ruby 3.2.6, failed with ssl error, tried again with `--with-openssl-dir` and it worked.

**Comment by monfresh** [comment]
Understood. I wanted to point out a common point of confusion where people try to use `--with-openssl-dir` (or some other method of pointing to OpenSSL) that points to OpenSSL 3, while trying to install Ruby versions older than 3.1, which don't support OpenSSL 3, and will cause the Ruby compilation to fail.

**Comment by crpahl** [comment]
Yeah exactly what @monfresh [shared](https://github.com/rvm/rvm/issues/5380#issuecomment-2452107484) is what we ran into for new people setting up their dev environments. We've migrated to `asdf` for this reason. Personally, I suggest everyone should just migrate to a new Ruby version manager at this point. `rvm` was amazing for us, but it seems to be abandoned at this point?
